### ISSUE TITLE
GRPOTrainer Is Extremely Slow on Small Dataset (5MB)

### ISSUE THREAD

**Issue by zahraajamali** [huggingface/trl#3574]
I am using the GRPOTrainer from HuggingFace's trl library with a small dataset (~5MB JSONL with precomputed rewards). Despite model quantization, LoRA, and precomputed rewards, training is extremely slow — taking **~5 minutes for just 5 steps out of 5001.**

System Configuration
GPU: NVIDIA Quadro RTX 6000 (24 GB VRAM)
Datasets size: 5 MB, ~2,000 samples

## GRPO Training Configuration

dataset = load_dataset("json", data_files={"train": "cleaned_train_with_rewards.jsonl"})
train_dataset = dataset["train"]

###  Load Model & Tokenizer (with LoRA + Quant) 
from model_loader import get_lora_config
lora_config = get_lora_config()

model, tokenizer = load_model_and_tokenizer(lora_config)
tokenizer.pad_token = tokenizer.eos_token

model.gradient_checkpointing_enable()
model.config.use_cache = False

### Reward Function 
reward_lookup = {
   ex["prompt"]: float(ex["delta_reward"])
   for ex in train_dataset
}

def compute_rewards(prompts, completions, **kwargs):
   return [reward_lookup.get(p, 0.0) for p in prompts]

### Training Args 
training_args = GRPOConfig(
   output_dir="grpo_sft_lora",
   logging_dir="./logs",
   logging_steps=50,
   report_to="none",  
   per_device_train_batch_size=4,
   gradient_accumulation_steps=2,
   save_strategy="epoch",
   learning_rate=2e-5,
   num_train_epochs=3,
   bf16=True,
   fp16=False,
   disable_tqdm=False
)

### GRPO Trainer 
trainer = GRPOTrainer(
   model=model,
   processing_class=tokenizer,
   args=training_args,
   train_dataset=train_dataset,
   reward_funcs=compute_rewards,
)

----------

- Using mistralai/Mistral-7B-Instruct-v0.2 model

- 4-bit quantization (bnb nf4) + LoRA

- Gradient checkpointing enabled

-Precomputed delta_reward is used from JSONL (no live reward generation)

# Question
Is there something I’m missing in the GRPOTrainer usage/config?

Should I disable generation explicitly?

Are there known performance issues with GRPOTrainer even with precomputed rewards?

Any advice on reducing runtime or confirming the trainer is not doing unnecessary generation would be helpful.

**Comment by qgallouedec** [comment]
You should try using vLLM. Please refer to the doc: https://huggingface.co/docs/trl/en/speeding_up_training?vllm+examples=GRPO

**Comment by albertvillanova** [comment]
Hey @zahraajamali, sorry for the very slow follow-up here — catching up on the backlog. 

The slowness isn't a bug: GRPO has to *generate* completions from the policy at every step (see https://github.com/huggingface/trl/blob/c3f21ede9afc0e2cadba183b3be376c5405cf65b/trl/trainer/grpo_trainer.py#L1818-L1874), and that is the dominant cost on a 7B model. Precomputing rewards only saves the (cheap) reward-evaluation step; it doesn't remove generation, and your lookup also returns the same reward for every completion of a given prompt, which means GRPO sees zero advantage and learns nothing. As @qgallouedec already mentioned, vLLM is the right knob — see https://huggingface.co/docs/trl/en/speeding_up_training?vllm+examples=GRPO. 

Closing as not-a-bug, but happy to re-open if you hit something more specific with a minimal repro and `trl env` output.
